### PR TITLE
feat: support drop partition table

### PR DESCRIPTION
**Pull request by chunshao90**
# Which issue does this PR close?

Closes #

# Rationale for this change
 Support drop partition table.
 

# What changes are included in this PR?
* Impl DropPartitionTableProcedure.
* Add TestCreateAndDropPartitionTable.

# Are there any user-facing changes?
Support drop partition table.

# How does this change test
Add new test.

### CODE REVIEW

File: server/coordinator/procedure/drop_partition_table.go [hunk 1]
```diff
@@ -0,0 +1,335 @@
...
+)
+
+// fsm state change:
+// ┌────────┐     ┌──────────────────────┐     ┌────────────────────┐     ┌───────────┐
+// │ Begin  ├─────▶     DropDataTable    ├─────▶ DropPartitionTable ├─────▶  Finish   │
+// └────────┘     └──────────────────────┘     └────────────────────┘     └───────────┘
+const (
+	eventDropDataTable            = "EventDropDataTable"
+	eventDropPartitionTable       = "EventDropPartitionTable"
+	eventDropPartitionTableFinish = "EventSuccess"
```
Comment (ZuLiangWang) [inline server/coordinator/procedure/drop_partition_table.go:28]: Make the variable value consistent with the name.

File: server/coordinator/procedure/drop_partition_table.go [hunk 2]
```diff
@@ -0,0 +1,335 @@
...
+}
+
+func (p *DropPartitionTableProcedure) convertToMeta() (Meta, error) {
+	p.lock.RLock()
+	defer p.lock.RUnlock()
+
+	rawData := dropPartitionTableRawData{
+		ID:       p.id,
+		FsmState: p.fsm.Current(),
+		State:    p.state,
```
Comment (ZuLiangWang) [inline server/coordinator/procedure/drop_partition_table.go:187]: `DropTableResult` is defined in `dropPartitionTableRawData` but not set.

File: server/coordinator/procedure/drop_partition_table.go [hunk 3]
```diff
@@ -0,0 +1,335 @@
...
+
+// 2. Drop partition table in target node.
+func dropPartitionTableCallback(event *fsm.Event) {
+	req, err := getRequestFromEvent[*dropPartitionTableCallbackRequest](event)
+	if err != nil {
+		cancelEventWithLog(event, err, "get request from event")
+		return
+	}
+
+	err = dropTable(event, req.sourceReq.Name)
```
Comment (ZuLiangWang) [inline server/coordinator/procedure/drop_partition_table.go:250]: Partition table maybe opened on multiple shards, but it has only been deleted once. The opened partition table needs to be closed by reopening the shard. Please refer to the logic of creating a table.

File: server/coordinator/procedure/drop_partition_table.go [hunk 4]
```diff
@@ -0,0 +1,335 @@
...
+const (
+	eventDropDataTable            = "EventDropDataTable"
+	eventDropPartitionTable       = "EventDropPartitionTable"
+	eventDropPartitionTableFinish = "EventSuccess"
+
+	stateDropPartitionTableBegin  = "StateBegin"
+	stateDropDataTable            = "StateDropDataTable"
+	stateDropPartitionTable       = "StateDropPartitionTable"
+	stateDropPartitionTableFinish = "StateFinish"
+)
```
Comment (ZuLiangWang) [inline server/coordinator/procedure/drop_partition_table.go:34]: Add close partition table event / state.

Comment (chunshao90) [reply server/coordinator/procedure/drop_partition_table.go:34]: `context` is more simpler.

File: server/coordinator/procedure/create_drop_partition_table_test.go [hunk 5]
```diff
@@ -0,0 +1,140 @@
...
+			checkTable(t, c, subTableName, true)
+		}
+	}
+
+	// Drop table.
+	for i := 0; i < testTableNum; i++ {
+		tableName := fmt.Sprintf("%s_%d", testTableName0, i)
+		subTableNames := genSubTables(tableName, testSubTableNum)
+		testDropPartitionTable(t, dispatch, c, s, tableName, subTableNames)
+	}
```
Comment (ZuLiangWang) [inline server/coordinator/procedure/create_drop_partition_table_test.go:49]: Add new line.

File: server/coordinator/procedure/factory.go [hunk 6]
```diff
@@ -60,6 +60,10 @@ type DropTableRequest struct {
 	OnFailed    func(error) error
 }
 
+func (d DropTableRequest) IsPartitionTable() bool {
+	return d.SourceReq.PartitionTableInfo != nil
```
Comment (ZuLiangWang) [inline server/coordinator/procedure/factory.go:64]: Check `partitionTableInfo.subTableNames`, you can refer to `CreateTableRequest`.

File: server/coordinator/procedure/drop_partition_table.go [hunk 7]
```diff
@@ -0,0 +1,421 @@
...
+// ┌────────┐     ┌────────────────┐     ┌────────────────────┐     ┌──────────────────────┐     ┌───────────┐
+// │ Begin  ├─────▶  DropDataTable ├─────▶ DropPartitionTable ├─────▶ ClosePartitionTables ├─────▶  Finish   │
+// └────────┘     └────────────────┘     └────────────────────┘     └──────────────────────┘     └───────────┘
+const (
+	eventDropDataTable            = "EventDropDataTable"
+	eventDropPartitionTable       = "EventDropPartitionTable"
+	eventClosePartitionTables     = "EventClosePartitionTables"
+	eventDropPartitionTableFinish = "eventDropPartitionTableFinish"
+
+	stateDropPartitionTableBegin  = "StateBegin"
```
Comment (ZuLiangWang) [inline server/coordinator/procedure/drop_partition_table.go:31]:
```suggestion
	stateDropPartitionTableBegin  = "StateDropPartitionTableBegin"
```

File: server/coordinator/procedure/drop_partition_table.go [hunk 8]
```diff
@@ -0,0 +1,421 @@
...
+
+// fsm state change:
+// ┌────────┐     ┌────────────────┐     ┌────────────────────┐     ┌──────────────────────┐     ┌───────────┐
+// │ Begin  ├─────▶  DropDataTable ├─────▶ DropPartitionTable ├─────▶ ClosePartitionTables ├─────▶  Finish   │
+// └────────┘     └────────────────┘     └────────────────────┘     └──────────────────────┘     └───────────┘
+const (
+	eventDropDataTable            = "EventDropDataTable"
+	eventDropPartitionTable       = "EventDropPartitionTable"
+	eventClosePartitionTables     = "EventClosePartitionTables"
+	eventDropPartitionTableFinish = "eventDropPartitionTableFinish"
```
Comment (ZuLiangWang) [inline server/coordinator/procedure/drop_partition_table.go:29]:
```suggestion
	eventDropPartitionTableFinish = "EventDropPartitionTableFinish"
```

File: server/coordinator/procedure/drop_partition_table.go [hunk 9]
```diff
@@ -0,0 +1,421 @@
...
+		return storage.Table{}, cluster.DropTableResult{}, errors.WithMessage(err, "get request from event")
+	}
+
+	table, exists, err := request.cluster.GetTable(request.schemaName(), tableName)
+	if err != nil {
+		return storage.Table{}, cluster.DropTableResult{}, errors.WithMessage(err, "cluster get table")
+	}
+	if !exists {
+		log.Warn("drop non-existing table", zap.String("schema", request.schemaName()), zap.String("table", tableName))
+		return storage.Table{}, cluster.DropTableResult{}, ErrTableNotExists
```
Comment (ZuLiangWang) [inline server/coordinator/procedure/drop_partition_table.go:374]: If return `ErrTableNotExists` when table not exists, then `dropPartitionTable` and `dropDataTable` are not idempotent functions.

File: server/coordinator/procedure/drop_partition_table.go [hunk 10]
```diff
@@ -0,0 +1,421 @@
...
+			}
+		case stateClosePartitionTables:
+			if err := p.persist(ctx); err != nil {
+				return errors.WithMessage(err, "drop partition table procedure persist")
+			}
+			if err := p.fsm.Event(eventDropPartitionTableFinish, dropPartitionTableRequest); err != nil {
+				p.updateStateWithLock(StateFailed)
+				return errors.WithMessagef(err, "drop partition table procedure close partition tables")
+			}
+			p.updateStateWithLock(StateFinished)
```
Comment (ZuLiangWang) [inline server/coordinator/procedure/drop_partition_table.go:151]: Remove this line.

File: server/coordinator/procedure/drop_partition_table.go [hunk 11]
```diff
@@ -0,0 +1,421 @@
...
+
+	for {
+		switch p.fsm.Current() {
+		case stateDropPartitionTableBegin:
+			if err := p.persist(ctx); err != nil {
+				return errors.WithMessage(err, "drop partition table procedure persist")
+			}
+			if err := p.fsm.Event(eventDropDataTable, dropPartitionTableRequest); err != nil {
+				p.updateStateWithLock(StateFailed)
+				return errors.WithMessagef(err, "drop partition table procedure")
```
Comment (ZuLiangWang) [inline server/coordinator/procedure/drop_partition_table.go:125]:
```suggestion
				return errors.WithMessage(err, "drop partition table procedure")
```
And check other places in this file.

File: server/coordinator/procedure/drop_partition_table.go [hunk 12]
```diff
@@ -0,0 +1,434 @@
...
+	cluster  *cluster.Cluster
+	dispatch eventdispatch.Dispatch
+	storage  Storage
+
+	onSucceeded func(cluster.TableInfo) error
+	onFailed    func(error) error
+
+	req *metaservicepb.DropTableRequest
+
+	lock  sync.RWMutex
```
Comment (ZuLiangWang) [inline server/coordinator/procedure/drop_partition_table.go:65]:
```suggestion
	// Protect the state.
	lock  sync.RWMutex
```

**Review by ZuLiangWang — APPROVED** [review]
LGTM

**Review by ZuLiangWang — APPROVED** [review]
LGTM

**Pull request merged**